File: tests/src/nl/t64/game/rpg/components/door/DoorTest.java
package nl.t64.game.rpg.components.door;

import nl.t64.game.rpg.GameTest;
import nl.t64.game.rpg.audio.AudioEvent;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;


class DoorTest extends GameTest {

    private DoorContainer doorContainer;

    @BeforeEach
    private void setup() {
        doorContainer = new DoorContainer();
    }

    @Test
    void whenDoorsAreCreated_ShouldHaveCorrectDataInside() {
        Door door = doorContainer.getDoor("door_simple_left");
        assertThat(door.getType()).isEqualTo(DoorType.SMALL);
        assertThat(door.getSpriteId()).isEqualTo("simple_left");
        assertThatExceptionOfType(NullPointerException.class).isThrownBy(door::getKeyId);
        assertThat(door.getAudio()).isEqualTo(AudioEvent.SE_SMALL_DOOR);
        assertThat(door.getWidth()).isEqualTo(48f);
        assertThat(door.getHeight()).isEqualTo(96f);

        assertThat(door.isLocked()).isFalse();
        assertThat(door.isClosed()).isTrue();
        door.open();
        assertThat(door.isClosed()).isFalse();
        door.close();
        assertThat(door.isClosed()).isTrue();
    }

    @Test
    void whenDoorsWithLocksAreCreated_ShouldHaveCorrectDataInside() {
        Door door = doorContainer.getDoor("door_gate_mysterious_tunnel");
        assertThat(door.getType()).isEqualTo(DoorType.GATE);
        assertThat(door.getSpriteId()).isEqualTo("gate_with_lock");
        assertThat(door.getKeyId()).contains("key_mysterious_tunnel");
        assertThat(door.getAudio()).isEqualTo(AudioEvent.SE_GATE);
        assertThat(door.getWidth()).isEqualTo(48f);
        assertThat(door.getHeight()).isEqualTo(96f);

        assertThat(door.isLocked()).isTrue();
        assertThat(door.isClosed()).isTrue();
        door.unlock();
        assertThat(door.isLocked()).isFalse();
        assertThat(door.isClosed()).isTrue();
    }

    @Test
    void testForCoverage_UnusedConstructorsByJackson() {
        Door door1 = new Door("LARGE", "large_round");
        Door door2 = new Door("GATE", "gate_with_lock", "key_mysterious_tunnel");
        assertThatExceptionOfType(NullPointerException.class).isThrownBy(door1::getKeyId);
        assertThat(door2.getKeyId()).contains("key_mysterious_tunnel");
    }

}

File: tests/src/nl/t64/game/rpg/components/door/DoorTest.kt
package nl.t64.game.rpg.components.door

import nl.t64.game.rpg.GameTest
import nl.t64.game.rpg.audio.AudioEvent
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test


internal class DoorTest : GameTest() {

    private lateinit var doorContainer: DoorContainer

    @BeforeEach
    private fun setup() {
        doorContainer = DoorContainer()
    }

    @Test
    fun whenDoorsAreCreated_ShouldHaveCorrectDataInside() {
        val door = doorContainer.getDoor("door_simple_left")
        assertThat(door.type).isEqualTo(DoorType.SMALL)
        assertThat(door.spriteId).isEqualTo("simple_left")
        assertThat(door.keyId).isNull()
        assertThat(door.audio).isEqualTo(AudioEvent.SE_SMALL_DOOR)
        assertThat(door.width).isEqualTo(48f)
        assertThat(door.height).isEqualTo(96f)

        assertThat(door.isLocked).isFalse
        assertThat(door.isClosed).isTrue
        door.open()
        assertThat(door.isClosed).isFalse
        door.close()
        assertThat(door.isClosed).isTrue
    }

    @Test
    fun whenDoorsWithLocksAreCreated_ShouldHaveCorrectDataInside() {
        val door = doorContainer.getDoor("door_gate_mysterious_tunnel")
        assertThat(door.type).isEqualTo(DoorType.GATE)
        assertThat(door.spriteId).isEqualTo("gate_with_lock")
        assertThat(door.keyId).contains("key_mysterious_tunnel")
        assertThat(door.audio).isEqualTo(AudioEvent.SE_METAL_GATE)
        assertThat(door.width).isEqualTo(48f)
        assertThat(door.height).isEqualTo(96f)

        assertThat(door.isLocked).isTrue
        assertThat(door.isClosed).isTrue
        door.unlock()
        assertThat(door.isLocked).isFalse
        assertThat(door.isClosed).isTrue
    }

    @Test
    fun testForCoverage_UnusedConstructorsByJackson() {
        val door1 = Door(DoorType.LARGE, "large_round")
        val door2 = Door(DoorType.GATE, "gate_with_lock", "key_mysterious_tunnel")
        assertThat(door1.keyId).isNull()
        assertThat(door2.keyId).contains("key_mysterious_tunnel")
    }

}
